### PR TITLE
redis v4 instrumentation

### PR DESCRIPTION
**Pull request by garbados**
## Proposed Release Notes

- Support automatic instrumentation of Redis v4

## Links

- Closes https://github.com/newrelic/node-newrelic/issues/995

### CODE REVIEW

**Comment by garbados** [comment]
Notable differences between v3 and v4 instrumentation:

- v3 instrumentation would nest subsequent calls as dependent segments, whereas the v4 instrumentation does not nest segments. This is because -- I think -- v3 uses callbacks such that function calls would nest simply because callbacks are contained in their parent context, whereas v4 using promises only thus avoiding this issue. As a result, the v4 test suite has been modified to find segments appropriately, though this difference will cause data to be represented differently in the NRO interface.
- Context binding in v4 is rough, as it is done by nesting the `shim.recordOperation` call within a patched client constructor. Thus there will be problems when using multiple Redis clients side-by-side, as commands would be instrumented multiple times. I am still trying to puzzle through a solution to this -- assists welcome!

**Comment by garbados** [comment]
From slack, more context:

> basically, we don't instrument the redis client directly. instead we instrument a command queue, which every redis command passes through. however the client and this queue are isolated from one another: a client has a queue and a queue has a client, but both are private variables for one another so we can't get one from the other, so i can't associate a given command with a particular client, so i can't attach instance variables like the host or port or db name to the active segment

**Comment by garbados** [comment]
This looks good to me! I left a comment with a concern but otherwise this looks good to me. I'm amenable to bringing this out of draft status now.

File: lib/instrumentation/@node-redis/client.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      const clientOptions = getRedisParams(client.options)
+      CMDS_TO_INSTRUMENT.forEach((cmd) => {
+        shim.recordOperation(client, cmd, function wrapCommand(shim, fn, fnName, args) {
+          const [key, value] = args
+          const parameters = Object.assign({}, clientOptions)
+          // If selecting a database, subsequent commands
+          // will be using said database, update the clientOptions
+          // but not the current parameters(feature parity with v3)
+          if (cmd.toLowerCase() === 'select') {
+            clientOptions.database_name = key
```
[the remark this answers is not in the document]
Comment (bizob2828) [reply lib/instrumentation/@node-redis/client.js:25]: no.  if you take a look at the comment the database for the current running command would not get changed, any subsequent ones would though

[the remark this answers is not in the document]
Comment (garbados) [reply lib/instrumentation/@node-redis/client.js:25]: I see, I see.

**Review by bizob2828 — APPROVED** [review]

**Pull request merged**